milli/src/external_documents_ids.rs

```diff
@@ -47,28 +48,40 @@ impl<'a> ExternalDocumentsIds<'a> {
         }
     }
 
-    pub fn delete_ids<A: AsRef<[u8]>>(&mut self, other: fst::Set<A>) -> fst::Result<()> {
-        let other = fst::Map::from(other.into_fst());
-        let union_op = self.soft.op().add(&other).r#union();
+    /// Delete each (external document id, internal docid) pair found in `other` from `self`, as well
+    /// the soft-deleted document ids.
+    ///
+    /// ## Arguments
+    /// - `other`: a map whose keys are external ids and whose values are roaring bitmaps of internal docids.
+    /// An external id will be deleted from `self` only if its associated internal docid is found in the
+    /// corresponding roaring bitmap from `other`, or if it is soft-deleted.
+    pub(crate) fn delete_ids(
+        &mut self,
+        other: FxHashMap<Vec<u8>, RoaringBitmap>,
+    ) -> fst::Result<()> {
+        let mut new_hard_builder = fst::MapBuilder::memory();
 
+        let union_op = self.hard.op().add(&self.soft).r#union();
         let mut iter = union_op.into_stream();
-        let mut new_soft_builder = fst::MapBuilder::memory();
         while let Some((external_id, docids)) = iter.next() {
-            if docids.iter().any(|v| v.index == 1) {
-                // If the `other` set returns a value here it means
-                // that it must be marked as deleted.
-                new_soft_builder.insert(external_id, DELETED_ID)?;
-            } else {
-                let value = docids.iter().find(|v| v.index == 0).unwrap().value;
-                new_soft_builder.insert(external_id, value)?;
+            // prefer selecting the ids from soft, always
+            let id = indexed_last_value(docids).unwrap();
+
+            if id != DELETED_ID
+                && !self.soft_deleted_docids.contains(id as u32)
+                && other.get(external_id).map_or(true, |b| !b.contains(id as u32))
+            {
+                new_hard_builder.insert(external_id, id)?;
```

Comment:
Thanks, this is way easier to follow, and if I follow correctly, should be faster :tada: